src/components/shared/TreeView.js

```diff
@@ -348,20 +361,35 @@ type TreeViewProps<DisplayData> = {|
 class TreeView<DisplayData: Object> extends React.PureComponent<
   TreeViewProps<DisplayData>
 > {
-  _specialItems: NodeIndex[];
+  _specialItems: Array<NodeIndex | void>;
   _visibleRows: NodeIndex[];
   _expandedNodes: Set<NodeIndex | null>;
   _list: VirtualList<NodeIndex> | null = null;
   _takeListRef = (list: VirtualList<NodeIndex> | null) => (this._list = list);
 
   constructor(props: TreeViewProps<DisplayData>) {
     super(props);
-    this._specialItems =
-      props.selectedNodeId === null ? [] : [props.selectedNodeId];
+
+    this._updateSpecialItems(props);
     this._expandedNodes = new Set(props.expandedNodeIds);
     this._visibleRows = this._getAllVisibleRows(props);
   }
 
+  _updateSpecialItems(props: TreeViewProps<DisplayData>) {
+    this._specialItems = [];
+
+    if (props.selectedNodeId !== null) {
+      this._specialItems[0] = props.selectedNodeId;
+    }
+
+    if (
+      props.rightClickedNodeId !== undefined &&
+      props.rightClickedNodeId !== null
+    ) {
+      this._specialItems[1] = props.rightClickedNodeId;
+    }
+  }
```

Comment:
I wonder if we should implement a memoized method instead of managing this local state.